File: examples/ios/SopranoDemo/Sources/TtsViewModel.swift
import Foundation
import os
import SopranoEngine

private let logger = Logger(subsystem: "com.example.soprano-demo", category: "TTS")

/// Thread-safe timing state shared between main actor and background threads.
/// Write ordering is safe: feedStartNanos is set before engine.feed(),
/// firstByteMs is set during writePcm, and read only after engine.drain() completes.
/// Lock ensures formal correctness regardless of ordering assumptions.
final class TimingState: @unchecked Sendable {
    private let lock = NSLock()
    private var _feedStartNanos: UInt64 = 0
    private var _firstByteMs: Int64 = -1

    var feedStartNanos: UInt64 {
        get { lock.withLock { _feedStartNanos } }
        set { lock.withLock { _feedStartNanos = newValue } }
    }

    var firstByteMs: Int64 {
        get { lock.withLock { _firstByteMs } }
        set { lock.withLock { _firstByteMs = newValue } }
    }

    func reset() {
        lock.withLock {
            _feedStartNanos = 0
            _firstByteMs = -1
        }
    }
}

@Observable
@MainActor
final class TtsViewModel {

    var isModelLoaded = false
    var isLoading = false
    var isSpeaking = false
    var error: String?
    var status = "Tap Load to start"
    var metrics = ""

    private var engine: SopranoTts?
    private var sink: AudioEngineSink?
    private var speakTask: Task<Void, Never>?
    private let _timing = TimingState()

    func loadModel(ep: String = "CPU") {
        guard !isLoading else { return }
        isLoading = true
        error = nil
        status = "Loading model..."
        metrics = ""

        // Release previous engine/sink before creating new ones (#6)
        sink?.release()
        sink = nil
        engine = nil

        Task.detached { [weak self] in
            guard let self else { return }

            do {
                let modelPath = Bundle.main.resourcePath! + "/Models"

                let newSink = AudioEngineSink(
                    onSentence: { index in
                        Task { @MainActor in
                            self.status = "Sentence \(index + 1) complete"
                        }
                    },
                    onComplete: {
                        Task { @MainActor in
                            self.isSpeaking = false
                            self.status = "Done"
                        }
                    },
                    onErr: { msg in
                        logger.error("Sink error: \(msg)")
                        Task { @MainActor in
                            self.isSpeaking = false
                            self.error = msg
                            self.status = "Error"
                        }
                    },
                    onFirstByte: {
                        let elapsed = DispatchTime.now().uptimeNanoseconds - self._timing.feedStartNanos
                        self._timing.firstByteMs = Int64(elapsed / 1_000_000)
                    }
                )

                let config = SopranoConfig(
                    modelPath: modelPath,
                    temperature: 0.0,
                    topK: 0,
                    topP: 0.95,
                    repetitionPenalty: 1.2,
                    executionProvider: ep == "CoreML" ? .coreMl : .cpu
                )

                logger.info("Loading model from \(modelPath) with EP: \(ep)")
                let t0 = DispatchTime.now().uptimeNanoseconds
                let newEngine = try SopranoTts(config: config, sink: newSink)
                let loadMs = (DispatchTime.now().uptimeNanoseconds - t0) / 1_000_000
                logger.info("Model loaded in \(loadMs)ms")

                await MainActor.run {
                    self.engine = newEngine
                    self.sink = newSink
                    self.isModelLoaded = true
                    self.isLoading = false
                    self.status = "Model loaded"
                    self.metrics = "Load: \(loadMs)ms"
                }
            } catch {
                logger.error("Model load failed: \(error.localizedDescription)")
                await MainActor.run {
                    self.isLoading = false
                    self.error = error.localizedDescription
                    self.status = "Failed to load model"
                }
            }
        }
    }

    func speak(text: String) {
        guard engine != nil, !isSpeaking else { return }
        isSpeaking = true
        error = nil
        status = "Synthesizing..."
        _timing.reset()
        sink?.resetForNewSynthesis()

        speakTask = Task.detached { [weak self] in
            guard let self, let engine = await self.engine else { return }

            do {
                self._timing.feedStartNanos = DispatchTime.now().uptimeNanoseconds
                let feedStart = DispatchTime.now().uptimeNanoseconds
                try engine.feed(text: text)
                engine.drain()
                let synthMs = (DispatchTime.now().uptimeNanoseconds - feedStart) / 1_000_000

                let totalBytes = await self.sink?.totalBytesWritten ?? 0
                let totalSamples = totalBytes / 2
                let audioDurationSec = Double(totalSamples) / 32_000.0

                let blockedMs = (await self.sink?.totalBlockedNanos ?? 0) / 1_000_000
                let inferenceMs = Int64(synthMs) - blockedMs
                let inferenceSec = Double(inferenceMs) / 1_000.0
                let rtf = inferenceSec > 0 ? audioDurationSec / inferenceSec : 0.0

                let firstByte = self._timing.firstByteMs

                await MainActor.run {
                    let loadLine = self.metrics.split(separator: "\n")
                        .first { $0.hasPrefix("Load:") }
                        .map(String.init) ?? ""

                    var lines: [String] = []
                    if !loadLine.isEmpty { lines.append(loadLine) }
                    if firstByte >= 0 { lines.append("First byte: \(firstByte)ms") }
                    lines.append("Inference: \(inferenceMs)ms (wall: \(synthMs)ms)")
                    if blockedMs > 0 {
                        lines.append("Audio blocked: \(blockedMs)ms")
                    }
                    lines.append(String(format: "Audio: %.2fs (%lld samples)", audioDurationSec, totalSamples))
                    lines.append(String(format: "RTF: %.2fx", rtf))
                    self.metrics = lines.joined(separator: "\n")
                    self.isSpeaking = false
                    self.status = "Done"
                    logger.info("Synthesis done — inference: \(inferenceMs)ms, RTF: \(String(format: "%.2f", rtf))x, first byte: \(firstByte)ms")
                }
            } catch {
                logger.error("Synthesis failed: \(error.localizedDescription)")
                await MainActor.run {
                    self.isSpeaking = false
                    self.error = error.localizedDescription
                    self.status = "Synthesis failed"
                }
            }
        }
    }

    func stop() {
        speakTask?.cancel()
        engine?.flush()
        // flush() causes the Rust side to return from blocked write,
        // which unblocks the semaphore naturally
        isSpeaking = false
        status = "Stopped"
    }

    func cleanup() {
        stop()
        engine = nil
        sink?.release()
        sink = nil
    }
}

File: examples/android/app/src/main/java/com/example/soprano/TtsViewModel.kt
package com.example.soprano

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import uniffi.soprano_ffi.ExecutionProvider
import uniffi.soprano_ffi.SopranoConfig
import uniffi.soprano_ffi.SopranoTts

data class TtsUiState(
    val isModelLoaded: Boolean = false,
    val isLoading: Boolean = false,
    val isSpeaking: Boolean = false,
    val error: String? = null,
    val status: String = "Enter model path and tap Load",
    val metrics: String = "",
)

class TtsViewModel : ViewModel() {

    private val _uiState = MutableStateFlow(TtsUiState())
    val uiState: StateFlow<TtsUiState> = _uiState

    private var engine: SopranoTts? = null
    private var sink: AudioTrackSink? = null
    private var speakJob: Job? = null

    @Volatile private var feedStartNanos: Long = 0
    @Volatile private var firstByteMs: Long = -1

    fun loadModel(modelPath: String, executionProvider: ExecutionProvider = ExecutionProvider.CPU) {
        if (_uiState.value.isLoading) return

        _uiState.update { it.copy(isLoading = true, error = null, status = "Loading model...", metrics = "") }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                engine?.close()
                sink?.release()

                val newSink = AudioTrackSink(
                    onStart = { tag, sampleOffset ->
                        Log.d("SopranoTts", "Sentence ${tag + 1u} audio starts at sample $sampleOffset")
                        _uiState.update { it.copy(status = "Speaking sentence ${tag + 1u}…") }
                    },
                    onComplete = {
                        _uiState.update { it.copy(isSpeaking = false, status = "Done") }
                    },
                    onErr = { msg ->
                        Log.e("SopranoTts", "Engine error: $msg")
                        _uiState.update { it.copy(isSpeaking = false, error = msg, status = "Error") }
                    },
                    onFirstByte = {
                        firstByteMs = (System.nanoTime() - feedStartNanos) / 1_000_000
                    },
                )

                val config = SopranoConfig(
                    modelPath = modelPath,
                    temperature = 0.0f,
                    topK = 0u,
                    topP = 0.95f,
                    repetitionPenalty = 1.2f,
                    executionProvider = executionProvider,
                )

                val t0 = System.nanoTime()
                val newEngine = SopranoTts(config, newSink)
                val loadMs = (System.nanoTime() - t0) / 1_000_000

                engine = newEngine
                sink = newSink

                _uiState.update {
                    it.copy(
                        isModelLoaded = true,
                        isLoading = false,
                        status = "Model loaded",
                        metrics = "Load: ${loadMs}ms",
                    )
                }
            } catch (e: Exception) {
                Log.e("SopranoTts", "Failed to load model", e)
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        error = e.message ?: "Unknown error",
                        status = "Failed to load model",
                    )
                }
            }
        }
    }

    fun speak(text: String) {
        val eng = engine ?: return
        if (_uiState.value.isSpeaking) return

        _uiState.update { it.copy(isSpeaking = true, error = null, status = "Synthesizing...") }

        // Reset tracking
        firstByteMs = -1
        sink?.resetForNewSynthesis()

        speakJob = viewModelScope.launch(Dispatchers.IO) {
            try {
                feedStartNanos = System.nanoTime()
                // Feed each sentence as its own tagged segment. The tag is echoed
                // back through onSentenceStart when that segment's audio begins,
                // so the UI can report which sentence is currently playing.
                splitIntoSentences(text).forEachIndexed { index, sentence ->
                    eng.feed(sentence, index.toULong())
                }
                eng.drain()
                val synthMs = (System.nanoTime() - feedStartNanos) / 1_000_000

                // Actual audio duration from bytes written: bytes / 2 (16-bit) / 32000 Hz
                val totalBytes = sink?.totalBytesWritten ?: 0L
                val totalSamples = totalBytes / 2
                val audioDurationSec = totalSamples.toDouble() / 32000.0

                // Subtract time blocked in AudioTrack.write() to get true inference time
                val writeBlockMs = (sink?.totalWriteNanos ?: 0L) / 1_000_000
                val inferenceMs = synthMs - writeBlockMs
                val inferenceSec = inferenceMs.toDouble() / 1000.0
                val rtf = if (inferenceSec > 0) audioDurationSec / inferenceSec else 0.0

                val loadPart = _uiState.value.metrics
                    .lineSequence()
                    .firstOrNull { it.startsWith("Load:") }
                    ?: ""

                _uiState.update {
                    it.copy(
                        metrics = buildString {
                            if (loadPart.isNotEmpty()) appendLine(loadPart)
                            if (firstByteMs >= 0) appendLine("First byte: ${firstByteMs}ms")
                            appendLine("Inference: ${inferenceMs}ms (wall: ${synthMs}ms)")
                            appendLine("Audio: %.2fs (%d samples)".format(audioDurationSec, totalSamples))
                            append("RTF: %.2fx".format(rtf))
                        },
                    )
                }
            } catch (e: Exception) {
                Log.e("SopranoTts", "Synthesis failed", e)
                _uiState.update {
                    it.copy(isSpeaking = false, error = e.message, status = "Synthesis failed")
                }
            }
        }
    }

    /** Split text into sentences on terminal punctuation, keeping the
     *  delimiter and dropping empty fragments. Falls back to the whole string
     *  if nothing splits, so a single feed always happens. */
    private fun splitIntoSentences(text: String): List<String> {
        val sentences = Regex("[^.!?]*[.!?]+|[^.!?]+\$")
            .findAll(text)
            .map { it.value.trim() }
            .filter { it.isNotEmpty() }
            .toList()
        return sentences.ifEmpty { listOf(text) }
    }

    fun stop() {
        engine?.flush()
        speakJob?.cancel()
        _uiState.update { it.copy(isSpeaking = false, status = "Stopped") }
    }

    override fun onCleared() {
        super.onCleared()
        engine?.close()
        sink?.release()
    }
}
